Add loop checks for other CO grids

maxbriel opened the pull request:
This adds checks for other CO grids and sets the binary state to `"FAILED"` instead of `"ERR"`, which seems to be used in `BinaryPopulation` as well when an error is thrown.

mkruckow commented:
On more thought: Currently, the checks assume a rectangular grid. This isn't the case at low m1 and low q. Should we already think about including a more detailed approach here to detect cases outside the grid?

mkruckow commented:
I think, we should always set the new state and event. (I'd suggest to clean the code by always having the same order of state and event code lines, which currently is sometimes one way some times the other way, which make is easy to accidentally swap the correct values those two should be set to.

maxbriel commented:
I'm not sure I understand your point on the rectangular grids.
Do you mean that in the HMS-HMS grid, m2 is forced to be larger than 0.5 in the grid creation and otherwise isn't created?
So depending on m1, q_min and m2_min is different?

This would already be a thing in the standard grid check. I would suggest to get `m2` from the interpolator and check if m1 and/or m2 is outside the those values.

As for the state+event change, I'm suggesting that I put a function in `utils/common_functions`, which sets the binary properties to "FAILED" and "ERR". That should provide the greatest flexibility for usage in multiple spaces (and files).

maxbriel commented:
During the dev meeting we had some discussion on handling FAILED binaries.

2 things for consideration: 
1. How we handle failed binaries.
2. The detail of error catching.

# 1. Handling failed binaries

Currently, a failed binary does not have a clear indicator in the dataframe/history that it failed. This binary failed, but it's unclear looking at the dataframe if it failed or just hasn't evolved further. Moreover, this leads to the issue with `event_f` being too long for storage in the output file, since these events/states are not meant to be a final state.

<img width="975" alt="Screenshot 2024-01-22 at 15 15 47" src="https://github.com/POSYDON-code/POSYDON/assets/14039563/718f956c-68a5-417c-bf99-3989de7d461e">

I suggest we use a uniform function, such as `set_binary_to_failed` to catch errors that we know about. This way, we know if we have binaries that are behaving outside what we expect. For example, we know that this example binary will fail because `S2_mass` is outside the grid for binary and single stars.

**Do we want to implement this everywhere?**

# 2. Detail of error catching.
This is more specific to the changes in this PR. There are some edge cases that @celiotine and @ka-rocha came across in Issue #208. While this issue was due to the initial sampling having a problem, we should catch in more detail if a binary is outside the grid.

A. star_mass outside the single star grid (addition in `step_detached`)
B. star_mass outside the binary grid (additions in `step_mesa` for each grid type)

The example binary above originally crashes with the following warning. 
<img width="636" alt="Screenshot 2024-01-22 at 15 31 31" src="https://github.com/POSYDON-code/POSYDON/assets/14039563/4ed913c1-9449-44a2-9118-ab32622664a9">

This is not informative. With this PR (due to additions A), they will crash like this:
<img width="752" alt="Screenshot 2024-01-22 at 15 33 59" src="https://github.com/POSYDON-code/POSYDON/assets/14039563/50fa33bf-cb10-4b6a-a8a5-a94a1fdd5834">

This provides more information for tracking what went wrong in the evolution.

For B. we now get these type of errors:
<img width="707" alt="Screenshot 2024-01-22 at 15 37 12" src="https://github.com/POSYDON-code/POSYDON/assets/14039563/dfb3f988-028f-46a8-8e30-2179e8d82c38">


<details><summary>B example binary</summary>
<p>

```

STAR1 = SingleStar(**{'mass': 13.889634,
                      'state': 'H-rich_Core_H_burning'})
STAR2 = SingleStar(**{'mass':0.490231,
                      'state': 'H-rich_Core_H_burning'})

BINARY = BinaryStar(STAR1, STAR2,  
                    **{'time': 0.0, 'state': 'detached', 'event': 'ZAMS', 'orbital_period':4513.150157, 'eccentricity': 0.0},
                    properties = sim_pop)

```


</p>
</details>

maxbriel commented:
@mkruckow would you be able to clarify on the rectangular grids assumption?

I believe that using `m2_min` and `m2_max` not `q` would not solve a secondary mass being outside the grid, since the first if-statement to determine if something is in the grid would already capture this binary.
To clarify, q_min =0.05, but for low mass primaries, m2*q >= 0.5 is enforced. Thus, the grid is not rectangular and some values won't be present at low m1?

It also seems to be the case that `m2_min/m1_min` from the interpolator gives `q_min=0.1`, but it's hard coded as 0.05.

mkruckow commented:
Usually, we have:
m1_min<=m1<=m1_max & q_min<=q<=q_max, where m1_min, m1_max, q_min=0.05 and q_max=1.0 are fixed values (those four values define a rectangular area).
But we use a third limiting criterion: 0.5=m2_min<=m2. Thus we effectively have one three constants: m1_min, m1_max and q_max. q_min is effectively a function of m1: f_q_min(m1)=max(q_min,m2_min/m1). This functional dependence turns the area to become a pentagon.

astroJeff commented:
Update (Jan 25): PR close to completion. @mkruckow and @maxbriel to finish up, resolving previous discussion about how to identify when binaries are outside the grids (grids are not exactly square)

maxbriel commented:
1. I've added a minimum period redirect, which follows the same direction as having a larger period. This ends up in `step_detached`, where it will be tagged as "initial_RLOF" and should stop. I don't think these systems go to the `step_merged`.
If this should be implemented, then that would be a new PR.

2. I've implemented a mass ratio selection as you suggested, based on $M_1$ and $q_\mathrm{min}$.

mkruckow commented:
1. OK, if the detached step can deal with those cases of too low period and don't redirect them to MESA RLO girds. Just for semantics: I'm not sure, that I like to get them classified as "initial_RLOF", when they come from a CO+star gird.
2. You have made the change at one place, please do it at all places, where `m2` or `q` is checked.

maxbriel commented:
For point 1, there's a slack conversation about this already, where it was noted that it felt a bit confusing.

For 2, I've changed the original code to also check calculate $q_\mathrn{min}$ for the HMS-HMS grids. The CO+star grids don't have this problem, because they're generated from CO and star masses that are always the same, right?
